src/Microsoft.SqlTools.ServiceLayer/FileBrowser/FileBrowserService.cs

```diff
@@ -0,0 +1,264 @@
+            }
+        }
+
+        internal async Task HandleFileBrowserValidateRequest(
+            FileBrowserValidateParams fileBrowserParams,
+            RequestContext<bool> requestContext)
+        {
+            try
+            {
+                Task.Run(() => RunFileBrowserValidateTask(fileBrowserParams));
+                await requestContext.SendResult(true);
+            }
+            catch
+            {
+                await requestContext.SendResult(false);
+            }
+        }
+
+        internal async Task HandleFileBrowserCloseRequest(
+            FileBrowserCloseParams fileBrowserParams,
+            RequestContext<FileBrowserCloseResponse> requestContext)
+        {
+            FileBrowserCloseResponse response = new FileBrowserCloseResponse();
+            if (this.ownerToFileBrowserMap.ContainsKey(fileBrowserParams.OwnerUri))
+            {
+                this.ownerToFileBrowserMap.Remove(fileBrowserParams.OwnerUri);
+                response.Succeeded = true;
+            }
+            await requestContext.SendResult(response);
+        }
+
+        #endregion
+
+        internal async Task RunFileBrowserOpenTask(FileBrowserOpenParams fileBrowserParams)
+        {
+            FileBrowserOpenCompleteParams result = new FileBrowserOpenCompleteParams();
+
+            try
+            {
+                ConnectionInfo connInfo;
+                this.ConnectionServiceInstance.TryFindConnection(
+                    fileBrowserParams.OwnerUri,
+                    out connInfo);
+                DbConnection dbConn = null;
+                connInfo.TryGetConnection(ConnectionType.Default, out dbConn);
+                SqlConnection conn = ReliableConnectionHelper.GetAsSqlConnection((IDbConnection)dbConn);
+
+                if (conn != null)
+                {
+                    FileBrowserOperation browser = new FileBrowserOperation(conn, fileBrowserParams.ExpandPath, fileBrowserParams.FileFilters);
+                    browser.PopulateFileTree();
+                    this.ownerToFileBrowserMap.Add(fileBrowserParams.OwnerUri, browser);
+                    result.FileTree = browser.FileTree;
+                    result.Succeeded = true;
+                }
+                else
+                {
+                    result.Succeeded = false;
+                }
+            }
+            catch (Exception ex)
+            {
+                result.Succeeded = false;
+                result.Message = ex.Message;
+            }
+
+            await ServiceHost.SendEvent(FileBrowserOpenCompleteNotification.Type, result);
+        }
+
+        internal async Task RunFileBrowserExpandTask(FileBrowserExpandParams fileBrowserParams)
+        {
+            FileBrowserExpandCompleteParams result = new FileBrowserExpandCompleteParams();
+            try
+            {
+                FileBrowserOperation browser = this.ownerToFileBrowserMap[fileBrowserParams.OwnerUri];
```

Comment:
>ownerToFileBrowserMap[fileBrowserParams.OwnerUri] [](start = 52, length = 49)

doesn't this throw if key doesn't exist?  I thought you needed to check the Dictionary with ContainsKey before accessing via the indexer?